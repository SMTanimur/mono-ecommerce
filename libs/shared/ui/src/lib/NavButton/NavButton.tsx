import React, { ReactNode, useMemo } from 'react';

import { Button } from '@ecommerce/shared/ui/Button';
import { ICONS, Icon } from '@ecommerce/shared/ui/Icon';

type TypeProp<T extends keyof typeof ICONS> = keyof Pick<typeof ICONS, T>;

export type NavButtonProps = (
  | {
      type: TypeProp<'user'>;
      loading: boolean;
    }
  | {
      type: TypeProp<'cartBlack' | 'heart'>;
      loading?: boolean;
    }
) & { children?: ReactNode };

export function getHref(type: NavButtonProps['type']): string {
  switch (type) {
    case 'cartBlack':
      return 'cart';
    case 'heart':
      return 'favorite';
    default:
      return type;
  }
}

export const NavButton: React.FC<NavButtonProps> = (props) => {
  const { type, loading, children } = props;

  const href = useMemo(() => getHref(type), [type]);

  return (
    <Button
      tag="a"
      type="transparent"
      elProps={{
        href: `/${href}`,
        title: `${href} page`,
      }}
      disabled={loading}
    >
      <Icon icon={type} loading={loading} />
      {children}
    </Button>
  );
};
