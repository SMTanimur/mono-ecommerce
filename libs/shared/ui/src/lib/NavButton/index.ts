import { NavButton, NavButtonProps } from './NavButton';

export { NavButton };

export type { NavButtonProps };
