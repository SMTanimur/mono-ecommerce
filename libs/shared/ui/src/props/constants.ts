import { ButtonProps } from '@ecommerce/shared/ui/Button';
import {CartItemProps} from "@ecommerce/shared/ui/CartItem";
import {ProductSidebarProps} from "@ecommerce/shared/ui/ProductSidebar";
import {CartSidebarProps} from "@ecommerce/shared/ui/CartSidebar";
import {CategoryCardProps} from "@ecommerce/shared/ui/CategoryCard";
import {IconProps} from "@ecommerce/shared/ui/Icon";
import {ItemCardProps} from "@ecommerce/shared/ui/ItemCard";
import {NavButtonProps} from "@ecommerce/shared/ui/NavButton";
import {ProductGalleryProps} from "@ecommerce/shared/ui/ProductGallery";

import cat from './cat.jpg';
import cat2 from './cat2.jpeg';

console.log(cat2)

export const PRODUCT: ProductSidebarProps['product'] = {
  slug: 'some-name',
  price: 65,
  name: 'Some name',
  images: {
    data: [
      {
        attributes: {
          url: cat,
        },
      },
      {
        attributes: {
          url: cat2,
        },
      },
    ],
  },
  smallDescription: 'Small Description',
  description: 'Description',
  information: [
    {
      title: 'Title1',
      content: 'content1',
    },
  ],
};

export const ITEM: ItemCardProps['item'] = {
  slug: 'cat-cat',
  price: 65,
  name: 'cat cat',
  images: {
    data: [
      {
        attributes: {
          url: cat,
        },
      },
      {
        attributes: {
          url: cat2,
        },
      },
    ],
  },
  smallDescription: 'smallDescription',
  description: 'description',
};

export const BUTTON_DEFAULT_PROPS: ButtonProps = {
  children: 'hello',
  tag: 'button',
  size: 'base',
  type: 'primary',
  round: 'full',
};

export const CART_ITEM_DEFAULT_PROPS: CartItemProps = {
  inCartAmount: 1,
  item: PRODUCT,
  handleCartMinusClick: () => {
    console.log('cart minus');
  },
  handleCartPlusClick: () => {
    console.log('cart plus');
  },
  handleRemoveFromCart: () => {
    console.log('cart remove');
  },
  toggleFavorite: () => {
    console.log('favorite');
  },
};

export const CART_SIDEBAR_DEFAULT_PROPS: CartSidebarProps = {
  cartTotal: 100,
  isUser: true,
  handleCheckoutClick: async () => {
    console.log('checkout');
  },
  isLoading: false,
};

export const CATEGORY_CARD_DEFAULT_PROPS: CategoryCardProps = {
  category: {
    smallDescription:
      'smalll des',
    image: {
      data: {
        attributes: {
          url: cat,
        },
      },
    },
    slug: '1',
  },
};
export const ICON_DEFAULT_PROPS: IconProps = {
  icon: 'cart',
};

export const ITEM_CARD_DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleCartClick: () => console.log('cart clicked'),
  handleAmountCartClick: () => console.log('amount cart clicked'),
  handleFavoriteClick: () => console.log('favorite clicked'),
};
export const NAV_BUTTON_DEFAULT_PROPS: NavButtonProps = {
  type: 'heart',
};
export const PRODUCT_GALLERY_DEFAULT_PROPS: ProductGalleryProps = {
  images: [
    {
      attributes: {
        url: cat,
      },
    },
    {
      attributes: {
        url: cat2,
      },
    },
  ],
};
