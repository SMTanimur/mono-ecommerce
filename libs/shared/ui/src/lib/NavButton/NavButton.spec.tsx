import '@testing-library/jest-dom';

import { RenderResult, render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';

import { NAV_BUTTON_DEFAULT_PROPS } from '@ecommerce/shared/props/constants';

import { NavButton, getHref } from './NavButton';

describe('NavButton', () => {
  let renderResult: RenderResult;
  const props = NAV_BUTTON_DEFAULT_PROPS;

  beforeEach(() => {
    renderResult = render(<NavButton {...props} />);
  });

  it('should render successfully', () => {
    const { baseElement } = renderResult;

    const tree = create(<NavButton {...props} />).toJSON();
    expect(tree).toMatchSnapshot();

    expect(baseElement).toBeTruthy();
  });

  it('proper href and title', () => {
    const href = getHref(props.type);

    const title = `${href} page`;
    const link = `/${href}`;

    const linkElement = screen.getByRole('link', { name: title });
    expect(linkElement).toHaveAttribute('href', link);
  });

  it('disabled if loading', () => {
    renderResult.rerender(<NavButton {...props} loading={true} />);

    const linkElement = screen.getByRole('link');

    expect(linkElement).toHaveClass('opacity-50 pointer-events-none');
  });
});
