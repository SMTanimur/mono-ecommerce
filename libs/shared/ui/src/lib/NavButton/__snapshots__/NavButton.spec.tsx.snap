// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavButton should render successfully 1`] = `
<a
  className="inline-block cursor-pointer transform-gpu transition-all active:scale-92 rounded-full p-3 hover:(bg-gray-100) active:(bg-gray-300)"
  href="/favorite"
  onClick={[Function]}
  onMouseEnter={[Function]}
  title="favorite page"
>
  <div
    aria-hidden="true"
    className="relative flex items-center select-none pointer-events-none h-5 w-5"
  >
    <svg
      className="feather feather-heart "
      fill="none"
      height={24}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width={24}
    >
      <g
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path d=\\"M20.84 4.61a5.5 5.5 0 0 0-7.78 0L12 5.67l-1.06-1.06a5.5 5.5 0 0 0-7.78 7.78l1.06 1.06L12 21.23l7.78-7.78 1.06-1.06a5.5 5.5 0 0 0 0-7.78z\\"></path>",
          }
        }
      />
    </svg>
  </div>
</a>
`;
